### PR TITLE
[feat] Make sanitize_repo report all found errors

### LINKED ISSUE #81: Sanitizer does not tell which file had an error
Instead of giving this:

```
[ERROR] PlugError: ['Line 36: Missing prefix', 'Line 57: Missing prefix', 'Line 78: Missing prefix', 'Line 91: Missing prefix', 'Line 104: Missing prefix', 'Line 126: Missing prefix', 'Line 144: Missing prefix', 'Line 157: Missing prefix', 'Line 168: Missing prefix', 'Line 209: Missing prefix']
```

We probably want it to tell us what file caused this as well, (they are all the same file, since sanitizer reports errors on a file to file basis)

### PR DESCRIPTION
**Pull request by tohanss**
Some things have been changed also, it is now important to check syntax yourself before sanitizing

Fix #81

### CODE REVIEW

File: repobee_sanitizer/_sanitize.py [hunk 1]
```diff
@@ -14,19 +14,17 @@
...
     """Runs the sanitization protocol on a given file. This can either remove
-    the file or give back a sanitized version
+    the file or give back a sanitized version. File must be syntax checked
+    before running this.
 
-    Args:
-        file_abs_path:
-            The absolute file path to the file you wish to sanitize.
+    Args: file_abs_path: The absolute file path to the file you wish to
+        sanitize.
```
Comment (slarse) [inline repobee_sanitizer/_sanitize.py:21]:
```suggestion
    Args:
        file_abs_path: The absolute file path to the file you wish to sanitize.
```

File: repobee_sanitizer/_sanitize.py [hunk 2]
```diff
@@ -14,19 +14,17 @@
...
 
-    Returns:
-        We return the sanitized output text, but only if the file
-        was not removed.
+    Returns: We return the sanitized output text, but only if the file was not
```
Comment (slarse) [inline repobee_sanitizer/_sanitize.py:23]:
```suggestion
    Returns:
        We return the sanitized output text, but only if the file
        was not removed.
```

File: repobee_sanitizer/_sanitize_file.py [hunk 3]
```diff
@@ -24,16 +24,27 @@ def _sanitize_file(
...
-            args: Parsed and processed args from the RepoBee CLI.
-            api: A platform API instance.
-        Returns:
-            A mapping (str -> List[plug.Result]) that RepoBee's CLI can use for
-            output.
+        Args: args: Parsed and processed args from the RepoBee CLI. api: A
+            platform API instance.
+
+        Returns: An optional plug.PlugResult if the syntax is invalid,
+            otherwise nothing.
```
Comment (slarse) [inline repobee_sanitizer/_sanitize_file.py:31]:
```suggestion
        Args:
            args: Parsed and processed args from the RepoBee CLI.
            api: A platform API instance.
        Returns:
            An optional plug.Result if the syntax is invalid, otherwise nothing.
```

File: repobee_sanitizer/_sanitize_repo.py [hunk 4]
```diff
@@ -42,10 +48,24 @@ def _sanitize_repo(
...
             LOGGER.info("Executing dry run")
             file_relpaths = _discover_dirty_files(args.repo_root)
-            _sanitize_files(args.repo_root, file_relpaths)
+            errors = _sanitize_files(args.repo_root, file_relpaths)
+            if errors:
+                return plug.Result(
+                    name="sanitize-repo",
+                    msg=_format_error_string(errors),
+                    status=plug.Status.ERROR,
+                )
```
Comment (slarse) [inline repobee_sanitizer/_sanitize_repo.py:57]: This is duplicated here and in the else branch. Put it after the entire if-statement instead.
```suggestion
```

File: repobee_sanitizer/_sanitize_repo.py [hunk 5]
```diff
@@ -42,10 +48,24 @@ def _sanitize_repo(
...
-            _sanitize_to_target_branch(args.repo_root, args.target_branch)
+            errors = _sanitize_to_target_branch(
+                args.repo_root, args.target_branch
+            )
+            if errors:
+                return plug.Result(
+                    name="sanitize-repo",
+                    msg=_format_error_string(errors),
+                    status=plug.Status.ERROR,
+                )
```
Comment (slarse) [inline repobee_sanitizer/_sanitize_repo.py:69]:
```suggestion
        if errors:
            return plug.Result(
                name="sanitize-repo",
                msg=_format_error_string(errors),
                status=plug.Status.ERROR,
            )
```

File: repobee_sanitizer/_sanitize_repo.py [hunk 6]
```diff
@@ -42,10 +48,24 @@ def _sanitize_repo(
...
```
Comment (slarse) [inline repobee_sanitizer/_sanitize_repo.py:68]: Same as the above comment.
```suggestion
```

File: repobee_sanitizer/_sanitize_repo.py [hunk 7]
```diff
@@ -114,8 +134,22 @@ def _discover_dirty_files(
...
+
+    for relpath in file_relpaths:
+        file_abspath = basedir / str(relpath)
+
+        errors = _syntax.check_syntax(file_abspath.read_text().split("\n"))
+        if errors:
+            files_with_errors.append(_FilesWithErrors(relpath, errors))
+
+    if files_with_errors:
+        return files_with_errors
```
Comment (slarse) [inline repobee_sanitizer/_sanitize_repo.py:151]: Just return the list. If it's empty, there are no errors. As a general advice, avoid returning `None` if possible.

```suggestion
    return files_with_errors
```

File: repobee_sanitizer/_sanitize_repo.py [hunk 8]
```diff
@@ -114,8 +134,22 @@ def _discover_dirty_files(
 
 def _sanitize_files(
     basedir: pathlib.Path, file_relpaths: List[_fileutils.RelativePath]
-) -> None:
-    """Sanitize the provided files."""
+) -> Optional[_FilesWithErrors]:
```
Comment (slarse) [inline repobee_sanitizer/_sanitize_repo.py:137]:
```suggestion
) -> List[_FileWithErrors]:
```

File: repobee_sanitizer/_sanitize_repo.py [hunk 9]
```diff
@@ -25,6 +26,11 @@
 LOGGER = daiquiri.getLogger(__file__)
 
 
+_FilesWithErrors = collections.namedtuple(
+    "_FileWithErrors", "file_rel_path Error"
+)
```
Comment (slarse) [inline repobee_sanitizer/_sanitize_repo.py:31]: * Don't use uppercase for attribute names.
* There can be multiple errors per file, so the attribute should probably be pluralized.
* You should match the variable name with the namedtuple name for clarity.

```suggestion
_FileWithErrors = collections.namedtuple(
    "_FileWithErrors", "file_rel_path errors"
)
```

File: repobee_sanitizer/_sanitize_repo.py [hunk 10]
```diff
@@ -128,9 +162,28 @@ def _sanitize_files(
             LOGGER.info(f"Shredded file {relpath}")
 
 
+def _format_error_string(files_with_errors: [_FilesWithErrors]) -> str:
```
Comment (slarse) [inline repobee_sanitizer/_sanitize_repo.py:165]: I'm guessing this is a list?

```suggestion
def _format_error_string(files_with_errors: List[_FileWithErrors]) -> str:
```

File: repobee_sanitizer/_sanitize_file.py [hunk 11]
```diff
@@ -24,16 +24,27 @@ def _sanitize_file(
...
         """
+        errors = _syntax.check_syntax(args.infile.read_text().split("\n"))
+        if errors:
+            msg = []
+            for error in errors:
+                msg.append(f"Line {error.line}: {error.msg}\n")
+
```
Comment (slarse) [inline repobee_sanitizer/_sanitize_file.py:38]: You have an error formatting function in `_sanitize_repo`. Make that a public function in a new module and reuse it instead of doing the same thing again here.

File: repobee_sanitizer/_sanitize_repo.py [hunk 12]
```diff
@@ -128,9 +162,28 @@ def _sanitize_files(
...
+                f"    {linestr if error.line else ''}{error.msg}"
+            )
+
+    return "\n".join(formated_str)
+
+
 def _sanitize_to_target_branch(
     repo_path: pathlib.Path, target_branch: str,
-) -> None:
+) -> Optional[_FilesWithErrors]:
```
Comment (slarse) [inline repobee_sanitizer/_sanitize_repo.py:186]: This hint is wrong (should be an optional _list_), but I would suggest removing the optional and just always returning a list.

```suggestion
) -> List[_FileWithErrors]:
```

Comment (slarse) [reply repobee_sanitizer/_sanitize_repo.py:186]: Just put `return []` at the end of the function.

File: repobee_sanitizer/_syntax.py [hunk 13]
```diff
@@ -62,55 +65,50 @@ def check_syntax(lines: List[str]) -> None:
...
 
         if (
             last == Markers.REPLACE or marker == Markers.END
         ) and not line.startswith(prefix):
-            errors.append(f"Line {line_number}: Missing prefix")
+            _error(f"Missing prefix: {prefix}")
 
     if last != Markers.END:
-        errors.append("Final block must be an END block")
+        errors.append(Error(None, "Global: Final block must be an END block"))
```
Comment (slarse) [inline repobee_sanitizer/_syntax.py:106]: The `Global: ` part should be handled in the output formatting, now you are repeating it here in multiple places. In the output formatting, if you encounter a line that's `None`, put `Global` instead of `Line x`.

File: tests/helpers/testhelpers.py [hunk 14]
```diff
@@ -45,7 +45,7 @@ def generate_valid_test_cases():
 def generate_invalid_test_cases():
     invalid_files = list(INVALID_CASES_BASEDIR.glob("*.in"))
     ids = [str(f.relative_to(INVALID_CASES_BASEDIR)) for f in invalid_files]
-    return [f.read_text(encoding="utf8") for f in invalid_files], ids
+    return [f.absolute() for f in invalid_files], ids
```
Comment (slarse) [inline tests/helpers/testhelpers.py:48]: Why is this changed to `f.absolute()`? Why not just return the text?

File: tests/test_sanitze.py [hunk 15]
```diff
@@ -18,11 +17,10 @@
...
 
 @pytest.mark.parametrize(
-    "inp", INVALID_TEST_CASE_ARGS, ids=INVALID_TEST_CASE_IDS
+    "abs_path", INVALID_TEST_CASE_ARGS, ids=INVALID_TEST_CASE_IDS
 )
-def test_sanitize_invalid(inp: str):
-    with pytest.raises(plug.PlugError):
-        _sanitize.sanitize_text(inp)
+def test_sanitize_invalid(abs_path: pathlib.Path):
+    assert _syntax.check_syntax(abs_path.read_text().split("\n"))
```
Comment (slarse) [inline tests/test_sanitze.py:23]: This is the only place I can find that you've changed in regards to returning the absolute path instead of the text. And the path is not used for anything but reading the text. Why not do that in the helper function instead?

You've also missed specifying `utf8` as the encoding when reading the text, so this will crash on a system that uses a different default encoding.

File: repobee_sanitizer/_sanitize_repo.py (hunk 7 above)
Comment (tohanss) [reply repobee_sanitizer/_sanitize_repo.py:151]: If we remove the if statement the code will not go on to the next loop

File: repobee_sanitizer/_sanitize_file.py (hunk 11 above)
Comment (tohanss) [reply repobee_sanitizer/_sanitize_file.py:38]: It has its own because I thought reporting how many files have an error is silly when you sanitize 1 file. Also telling the user what file is also silly since they had to specify that file themselves.

Comment (tohanss) [reply repobee_sanitizer/_sanitize_file.py:38]: So the change feels unnecessary for the work of creating a new module. Also if we're going to change how we report errors in the future (which we have talked about), we can do that then

Comment (slarse) [reply repobee_sanitizer/_sanitize_file.py:38]: So what happens e.g. with global errors in this single-file error formatting? Is that handled?

Comment (slarse) [reply repobee_sanitizer/_sanitize_file.py:38]: Moving one function that is used in one place to another module is something that takes roughly 10 seconds, so I don't see how it's a lot of work. I would strongly advice you to do this refactor, because not only does this duplicate the error output formatting from sanitize repo, it does so in a slightly more buggy way.

File: tests/helpers/testhelpers.py (hunk 14 above)
Comment (tohanss) [reply tests/helpers/testhelpers.py:48]: Before i removed the sanitize check from sanitize file, I think I used that, and then after removing the check i just fixed it, without realizing it was best to revert to how it was before, i will revert it.

File: repobee_sanitizer/_sanitize_file.py (hunk 11 above)
Comment (tohanss) [reply repobee_sanitizer/_sanitize_file.py:38]: Reasonable, ill change it

Comment (slarse) [reply repobee_sanitizer/_sanitize_file.py:38]: Cool. As you've probably noticed by now, nice output formatting is far from trivial. Corner cases abound, and it's really quite difficult to get it "just right". Therefore, it's something you really don't want to do twice :)

File: repobee_sanitizer/_syntax.py [hunk 16]
```diff
@@ -62,55 +61,50 @@ def check_syntax(lines: List[str]) -> None:
...
-                    "END block must follow START or REPLACE block"
-                )
+                _error("END block must follow START or REPLACE block")
             last = Markers.END
 
         if (
             last == Markers.REPLACE or marker == Markers.END
         ) and not line.startswith(prefix):
-            errors.append(f"Line {line_number}: Missing prefix")
+            _error(f"Missing prefix: {prefix}")
```
Comment (slarse) [inline repobee_sanitizer/_syntax.py:99]: good

File: tests/test_ext_commands.py [hunk 17]
```diff
@@ -44,20 +45,38 @@ def test_sanitize_valid_file(self, sanitizer_config, tmpdir):
         assert outfile.read_text(encoding="utf8") == out_text
 
     def test_removes_file_with_shred_marker(self, sanitizer_config, fake_repo):
-        """Test that sanitize-repo does not send any files that contain a shred
+        """Test that sanitize-file does not send any files that contain a shred
```
Comment (slarse) [inline tests/test_ext_commands.py:48]: This comment doesn't make sense, sanitize-file has no target branch. This is not what this test actually tests.

File: tests/test_sanitze.py [hunk 18]
```diff
@@ -18,11 +16,10 @@
 
 
 @pytest.mark.parametrize(
-    "inp", INVALID_TEST_CASE_ARGS, ids=INVALID_TEST_CASE_IDS
+    "text", INVALID_TEST_CASE_ARGS, ids=INVALID_TEST_CASE_IDS
 )
-def test_sanitize_invalid(inp: str):
-    with pytest.raises(plug.PlugError):
-        _sanitize.sanitize_text(inp)
+def test_sanitize_invalid(text: [str]):
```
Comment (slarse) [inline tests/test_sanitze.py:21]: You have a bad habit of putting type hints in brackets :)
```suggestion
def test_sanitize_invalid(text: str):
```
